truss/tests/test_control_truss_patching.py

```diff
@@ -0,0 +1,310 @@
+    verify_system_package_installed_on_container,
+    verify_system_requirement_not_installed_on_container,
+)
+from truss.truss_handle import TrussHandle
+
+
+def current_num_docker_images(th: TrussHandle) -> int:
+    return len(th.get_all_docker_images())
+
+
+@pytest.mark.integration
+@pytest.mark.parametrize(
+    "binary, python_version",
+    [
+        (binary, python_version)
+        for binary in [True, False]
+        for python_version in ["3.8", "3.9"]
+    ],
+)
+def test_control_truss_local_update_flow(binary, python_version, custom_model_control):
+    th = TrussHandle(custom_model_control)
+    th.update_python_version(python_version)
+    tag = "test-docker-custom-model-control-tag:0.0.1"
+
+    def predict_with_updated_model_code():
+        new_model_code = """
+class Model:
+    def predict(self, model_input):
+        return [2 for i in model_input]
+"""
+        model_code_file_path = custom_model_control / "model" / "model.py"
+        with model_code_file_path.open("w") as model_code_file:
+            model_code_file.write(new_model_code)
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_added_empty_directory():
+        # Adding empty directory should work
+        (custom_model_control / "model" / "dir").mkdir()
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_unpatchable_change():
+        # Changes that are not expressible with patch should also work
+        # Changes to data dir are not currently patch expressible
+        (custom_model_control / "data" / "dummy").touch()
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_python_requirement_added(req: str):
+        th.add_python_requirement(req)
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_python_requirement_removed(req):
+        th.remove_python_requirement(req)
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_system_requirement_added(pkg):
+        th.add_system_package(pkg)
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_system_requirement_removed(pkg):
+        th.remove_system_package(pkg)
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_ignored_changes():
+        top_pycache_path = custom_model_control / "__pycache__"
+        top_pycache_path.mkdir()
+        (top_pycache_path / "bla.pyc").touch()
+        model_pycache_path = custom_model_control / "model" / "__pycache__"
+        model_pycache_path.mkdir()
+        (model_pycache_path / "foo.pyc").touch()
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    with ensure_kill_all():
+        result = th.docker_predict([1], tag=tag, binary=binary)
+        assert result[0] == 1
+        orig_num_truss_images = len(th.get_all_docker_images())
+
+        result = predict_with_updated_model_code()
```

Comment:
does it makes sense to move these to smaller unit-like tests so that we indenitfy regressions one by one instead of having to debug the integration test every time?

Reply:
yes – I just copied over the existing test, let me make them more effective